### PR TITLE
Issue 2323 Implement versioning in controller metadata

### PR DESCRIPTION
**Pull request by shiveshr**
**Change log description**
1. Use VersionedSerializer to serialize/deserialize all controller metadata which adds versioning to the metadata. 
2. Controller tables (segment and history tables) can now have rows with different versions in the same table. This breaks several  properties of the table like fixed row sizes, O(1) navigation etc. This has been handled by adding new indexes per table that index start of each row in the table.  
3. Since we cannot perform atomic updates to multiple records like table and corresponding index, it has been handled during these metadata updates. This primarily impacts scale workflow because scale is where we update these tables.   

**Purpose of the change**
Implements versioning in controller metadata

**What the code does**
1. Serialization using VersionedSerliazer implemented for controller metadata. 
2. Scale workflow had additional complexity to determine if there was an ongoing scale by looking at records in different tables and comparing it with scale input. Prior to this change we had to deal with complexity arising from two table update. However, with indexes introduced, the complexity gets multiplied as we can be at different stages of update.  

So to handle such scenarios we have added a new transient node called `EpochTransitionRecord`. This node is created at the start of a new scale operation and removed when scale completes. This node simplifies the barrier logic for multiple scale operations which earlier relied of looking at the state in segment and history tables. 
All steps of scale operation retrieve this record and compare the state locally with this record to perform the step in idempotent fashion. 
We first update the index followed by updating the table. If we crash after updating the  index, then the index which is ahead  of table is erased. So indexes are no longer append only. But they only overwrite last set of records that have not been reflected in the corresponding table. 
We need to erase this index because subsequent attempt to compute table records may use a different version of the code whereby rendering index incorrect. 

All stream-store queries are updated to handle the indexes too.
 
3. refactored code --> renamed tables package to records as it keeps all stream store records.
 
**How to verify it**
New unit tests added and existing unit tests modified.

### CODE REVIEW

**Comment by shiveshr** [comment]
[Use custom serialization with version on controller metadata](https://trello.com/c/rmd8xTsI/21-use-custom-serialization-with-version-on-controller-metadata)

**Comment by shiveshr** [comment]
[Use custom serialization with version on controller metadata](https://trello.com/c/rmd8xTsI/21-use-custom-serialization-with-version-on-controller-metadata)

File: controller/src/main/java/io/pravega/controller/store/stream/records/ActiveTxnRecord.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            log.error("deserialization error for active txn record {}", e);
+            throw Lombok.sneakyThrow(e);
+        }
+        return activeTxnRecord;
+    }
+
+    public byte[] toByteArray() {
+        byte[] array;
+        try {
+            array = SERIALIZER.serialize(this).array();
```
Comment (andreipaduroiu) [inline controller/src/main/java/io/pravega/controller/store/stream/records/ActiveTxnRecord.java]: This is wrong. `serialize` returns an `ArrayView` which is a sub-segment of an array. If you look at `ArrayView`'s Javadoc, you'll see that `array()` returns the whole base array, which contains more than what you want.

You can just call `return SERIALIZER.serializer(this).getCopy()` which will handle the dirty work of copying for you.

Alternatively, if wherever you need this data accepts an `InputStream`, you can call `ArrayView.getReader()` which will help you avoid creating the intermediate buffer.

Comment (andreipaduroiu) [reply controller/src/main/java/io/pravega/controller/store/stream/records/ActiveTxnRecord.java]: Please fix this everywhere.

Comment (shiveshr) [reply controller/src/main/java/io/pravega/controller/store/stream/records/ActiveTxnRecord.java]: Fixed

**Pull request closed without merging**